pkg/cloudprovider/provider/openstack/helper.go

```diff
@@ -324,21 +331,65 @@ func getFreeFloatingIPs(client *gophercloud.ProviderClient, region string, float
 	return freeFIPs, nil
 }
 
-func assignFloatingIP(client *gophercloud.ProviderClient, region, ipID, instanceID, networkID string) error {
+func assignFloatingIPToInstance(client *gophercloud.ProviderClient, instanceID, floatingIPPoolName, region string, network *osnetworks.Network) error {
+	port, err := getInstancePort(client, region, instanceID, network.ID)
+	if err != nil {
+		return fmt.Errorf("failed to get instance port for network %s in region %s: %v", network.ID, region, err)
+	}
+
 	netClient, err := goopenstack.NewNetworkV2(client, gophercloud.EndpointOpts{Region: region})
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to create the networkv2 client for region %s: %v", region, err)
 	}
 
-	port, err := getInstancePort(client, region, instanceID, networkID)
+	floatingIPPool, err := getNetwork(client, region, floatingIPPoolName)
 	if err != nil {
-		return err
+		return osErrorToTerminalError(err, fmt.Sprintf("failed to get floating ip pool %q", floatingIPPoolName))
 	}
-	_, err = osfloatingips.Update(netClient, ipID, osfloatingips.UpdateOpts{
+
+	// We're only interested in the part which is vulnerable to concurrent access
+	started := time.Now()
+	glog.V(2).Infof("Assigning a floating IP to instance %s", instanceID)
+
+	floatingIPAssignLock.Lock()
+	defer floatingIPAssignLock.Unlock()
+
+	freeFloatingIps, err := getFreeFloatingIPs(client, region, floatingIPPool)
+	if err != nil {
+		return osErrorToTerminalError(err, "failed to get free floating ips")
+	}
+
+	var ip *osfloatingips.FloatingIP
+	if len(freeFloatingIps) < 1 {
+		if ip, err = createFloatingIP(client, region, floatingIPPool); err != nil {
+			return osErrorToTerminalError(err, "failed to allocate a floating ip")
+		}
+	} else {
+		ip = &freeFloatingIps[0]
+	}
+
+	_, err = osfloatingips.Update(netClient, ip.ID, osfloatingips.UpdateOpts{
 		PortID: &port.ID,
 	}).Extract()
+	if err != nil {
+		return fmt.Errorf("failed to update FloatingIP %s: %v", ip.ID, err)
+	}
+	secondsTook := time.Since(started).Seconds()
```

Comment:
Feels to me like this should be a metric?

Reply:
I though about it, but decided not introduce a metric here.
The values would depend on creating a new FloatingIP or just assigning.
And creating a metric for both parts individually, would mean creating a metric for each request to OpenStack effectively.

The time above is really just for informational purpose